crates/bevy_reflect/src/path.rs

```diff
@@ -104,270 +114,481 @@ impl<T: Reflect> GetPath for T {
-                        match current.reflect_mut() {
-                            ReflectMut::List(reflect_list) => {
-                                let list_index = value.parse::<usize>()?;
-                                let list_item = reflect_list.get_mut(list_index).ok_or(
-                                    ReflectPathError::InvalidListIndex {
-                                        index: current_index,
-                                        list_index,
-                                    },
-                                )?;
-                                current = list_item;
-                            }
-                            _ => {
-                                return Err(ReflectPathError::ExpectedStruct {
-                                    index: current_index,
-                                })
-                            }
-                        }
-                    } else {
-                        return Err(ReflectPathError::ExpectedIdent {
-                            index: current_index,
-                        });
-                    }
+        for (access, current_index) in PathParser::new(path) {
+            current = access?.read_field_mut(current, current_index)?;
+        }
+        Ok(current)
+    }
+}
+
+/// A path to a field within a type. Can be used like [`GetPath`] functions to get
+/// references to the inner fields of a type.
+#[derive(Clone, Debug, PartialEq, PartialOrd, Ord, Eq, Hash)]
+pub struct FieldPath(Box<[(Access, usize)]>);
+
+impl FieldPath {
+    /// Parses a [`FieldPath`] from a string. For the exact format, see [`GetPath`].
+    /// Returns an error if the string does not represent a valid path to a field.
+    pub fn parse(string: &str) -> Result<Self, ReflectPathError<'_>> {
+        let mut parts = Vec::new();
+        for (access, idx) in PathParser::new(string) {
+            parts.push((access?.to_owned(), idx));
+        }
+        Ok(Self(parts.into_boxed_slice()))
+    }
+
+    /// Gets a read-only reference of given field.
+    /// Returns an error if the path is invalid for the provided type.
+    pub fn field<'r, 'p>(
+        &'p self,
+        root: &'r dyn Reflect,
+    ) -> Result<&'r dyn Reflect, ReflectPathError<'p>> {
+        let mut current = root;
+        for (access, current_index) in self.0.iter() {
+            current = access.to_ref().read_field(current, *current_index)?;
+        }
+        Ok(current)
+    }
+
+    /// Gets a mutable reference of given field.
+    /// Returns an error if the path is invalid for the provided type.
+    pub fn field_mut<'r, 'p>(
+        &'p mut self,
+        root: &'r mut dyn Reflect,
+    ) -> Result<&'r mut dyn Reflect, ReflectPathError<'p>> {
+        let mut current = root;
+        for (access, current_index) in self.0.iter() {
+            current = access.to_ref().read_field_mut(current, *current_index)?;
+        }
+        Ok(current)
+    }
+
+    pub fn get_field<'r, 'p, T: Reflect>(
```

Comment:
This needs docs.

Reply:
This mirrors the same API for `GetPath`, if we update it here, we should update it in both places. The main difference is that one has a concrete type and the other returns a `dyn Reflect` reference.